### PR TITLE
packaging import fix

### PR DESCRIPTION
**Pull request by stefmdavis**
When trying to run a `git push` on any of our pipenv-cookiecutter repositories, both of the pipenv-setup pre-push hooks fail with the following error:
```
Traceback (most recent call last):
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/bin/pipenv-setup", line 5, in <module>
    from pipenv_setup.main import cmd
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/lib/python3.12/site-packages/pipenv_setup/main.py", line 25, in <module>
    from pipenv_setup import (
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/lib/python3.12/site-packages/pipenv_setup/lockfile_parser.py", line 3, in <module>
    from requirementslib import Lockfile, Requirement
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/lib/python3.12/site-packages/requirementslib/__init__.py", line 4, in <module>
    from .models.lockfile import Lockfile
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/lib/python3.12/site-packages/requirementslib/models/lockfile.py", line 13, in <module>
    from .requirements import Requirement
  File "/Users/stefmd/.cache/pre-commit/repoa7rtv7ax/py_env-python3.12/lib/python3.12/site-packages/requirementslib/models/requirements.py", line 26, in <module>
    from pip._vendor.packaging.specifiers import (
ImportError: cannot import name 'LegacySpecifier' from 'pip._vendor.packaging.specifiers'
```

After some research, I discovered that this issue was due, at its very heart, to the fact that the [`packaging`](https://pypi.org/project/packaging/) lib deprecated its `LegacyVersion` and `LegacySpecifier` functionality (read about that [here](https://github.com/pypa/packaging/issues/321)) from version 22.0 onwards.

How this relates to this particular problem is twofold:
1. `packaging` is used within the repo [here](https://github.com/anmut-consulting/pipenv-setup/blob/76482d061b0d38b932dcea5d36a42e9b85e26481/pipenv_setup/inconsistency_checker.py#L8), but THIS IS NOT THE PROBLEM since we pin its version in the [pipfile](https://github.com/anmut-consulting/pipenv-setup/blob/76482d061b0d38b932dcea5d36a42e9b85e26481/Pipfile#L17) to one that is compatible with version 21.0 which still contains the now deprecated functionality.
2. `packaging` also just so happens to be sub-sub-dependency of pipenv-setup via the [`requirementslib`](https://pypi.org/project/requirementslib/) package which depends on [`pip`](https://pypi.org/project/pip/) which then itself depends on `packaging`. THIS IS THE ROOT CAUSE OF THE PROBLEM.

Taking a dive down into this particular rabbit-hole, raises yet another twofold problem:
- even the latest versions of `requirementslib` still depends on the `LegacySpecifier` functionality from `packaging` via `pip`, as can be seen in the GitHub source code [here](https://github.com/sarugaku/requirementslib/blob/b19383c0189baec9827efd00f4dd6c95b20f74b3/src/requirementslib/models/requirements.py#L40) and the raised issue [here](https://github.com/sarugaku/requirementslib/issues/369).
- and crucially, `requirementslib` will install ANY version of `pip` greater than or equal to 22.2 which basically renders the package useless given it will by default install the latest version which has been updated to cater for the update in `packaging` and therefore give the error above.

In other words, `requirementslib` need to sort their sh*t out.

So in order to fix:
- I had to manually install `pip` and pin the version within the Pipfile to an older compatible version ([v24.0](https://github.com/pypa/pip/blob/24.0/src/pip/_vendor/vendor.txt)) which still depends on the older version of `packaging` containing the `LegacySpecifier` functionality ([v21.3](https://github.com/pypa/packaging/blob/b10b90a7d1c3afac6852462fa2548db231810adf/packaging/specifiers.py#L227)), as recommended [here](https://github.com/pypa/packaging/issues/530), as this will force `requirementslib` to use that version of `pip` which isn't problematic.

Problem solved. For now.

### CODE REVIEW

File: Pipfile
Comment (stefan-anmut) [inline Pipfile:1]: NOTE: most of the "changes" are just due to me reordering the packages according to alphabetical order.

**Review by jmartin-anmut — DISMISSED** [review]
I'm not a fan of adding `pip` to the `Pipfile` because of compatibility issues it might raise down the line, especially as `pipenv` uses its own version of `pip`... but in this scenario I don't see much choice. The only other option would be remove our dependency on `requirementslib`, but to be honest I don't know if there are other packages that would give us the same functionality, and might be more work than it's worth instead of just waiting for the issue with `requirementslib` to be fixed. It will only be a matter of time as other projects have done the same.

If we go ahead with this, we need to keep an eye on `requirementslib` so that when the issue is resolved, we can swiftly remove `pip` from the `Pipfile`. Maybe we create a separate GH issue for it so we don't forget.

Nice job figuring this one out!

**Comment by stefan-anmut** [comment]
> I'm not a fan of adding `pip` to the `Pipfile` because of other compatibility issues it might raise down the line

That's a very good point James. Luckily we have all dependencies locked down to a certain version in the Pipfile so, as long as it works now (which it does), I'm pretty sure it will continue to work. Until the day we decide to start updating things, but we'll cross that bridge when we get to it.

And yep, also good point on raising an issue to address this in the future. I've created one as suggested ([link](https://github.com/anmut-consulting/pipenv-setup/issues/8)).

Thanks for the review.

**Comment by csut-anmut** [comment]
The changes themselves look fine to me, although I'm cautiously apprehensive about pinning `pip` as this could cause us compatibility issues later (in other projects where pipenv-setup is installed as a hook or dependency, for example).

If I'm understanding correctly, the reason `pip` needs to be pinned is because the latest version of `pip` expects a fully updated version of `packaging`, which we can't use as `requirementslib` hasn't been updated to not use the now-deprecated `LegacyVersion` and `LegacySpecifier` functionality, right?

I echo James' sentiment that we should keep an eye on `requirementslib` and unpin `pip` if and when the issue is fixed.

**Comment by stefan-anmut** [comment]
Yes pretty much @csut-anmut.

But for the avoidance of any doubt, I'll try to explain it again:

The reason we need to pin `pip` is because `requirementslib` still uses old/deprecated `packaging` functionality via it's `pip` dependency (as can be seen from this import line in requirementslib [here](https://github.com/sarugaku/requirementslib/blob/591f194bb67d7b4124483ca62fdda8cd19088a86/src/requirementslib/models/requirements.py#L40)) BUT also specifies to import any version of pip >= 23.1 (as can be seen [here](https://github.com/sarugaku/requirementslib/blob/591f194bb67d7b4124483ca62fdda8cd19088a86/setup.cfg#L46)).

This is problematic because any version of pip > 24.0 will have removed this old/deprecated functionality by upgrading its own dependency on `packaging`.

So to fix, we have to force `requirementslib` to use a version of pip that still works (ie. 24.0), instead of the latest which it installs by default, by installing it and pinning its version in our Pipfile.

Hope that makes sense.

**Review by csut-anmut — DISMISSED** [review]
Thanks for the explanation. Approved 👍.

**Review by csut-anmut — APPROVED** [review]

**Review by jmartin-anmut — COMMENTED** [review]
looks good to me 👍

**Review by jmartin-anmut — APPROVED** [review]
